File: src/heap/furthest-building-you-can-reach.ts
// DIFFICULTY: Medium
//
// You are given an integer array heights representing the heights of buildings, some bricks, and some ladders.
//
// You start your journey from building 0 and move to the next building by possibly using bricks or ladders.
//
// While moving from building i to building i+1 (0-indexed),
//
// - If the current building's height is greater than or equal to the next building's height, you do not need a ladder
//   or bricks.
// - If the current building's height is less than the next building's height, you can either use one ladder or
//   (h[i+1] - h[i]) bricks.
//
// Return the furthest building index (0-indexed) you can reach if you use the given ladders and bricks optimally.
//
// See {@link https://leetcode.com/problems/furthest-building-you-can-reach/}
import { MinPriorityQueue } from '@datastructures-js/priority-queue';
export { furthestBuilding };

// SOLUTION:
//
// This cannot be solved with the sliding window technique; usually that technique involves finding a fixed size
// sub-string sub-array with constraints.  Here we need to dynamically adjust our resource (brick/ladder) usage at
// each step, making a greedy algorithm better to solve the problem.
//
// COMPLEXITY:
//
// We iterate through n heights in O(n).  Each iteration, we may push onto the heap, which is O(log(n)).  This results in
// O(n * log(n)) time complexity.
function furthestBuilding(heights: number[], bricks: number, ladders: number): number {
  // Use a heap/priority queue to store the number of jumps we need to make with either bricks or ladders.
  const heap = new MinPriorityQueue<number>();

  for (let i = 0; i < heights.length - 1; i++) {
    const delta = heights[i + 1] - heights[i];

    // If the height is same level or lower, we can jump across with no penalty.
    if (delta <= 0) {
      continue;
    }

    // If the height is higher by some amount, push it onto the heap and make a determination on whether we use a
    // ladder or some bricks to scale the difference.
    //
    // We should use ladders on the biggest deltas, so if we greedily consume the smallest deltas from the heap, as
    // long as we have enough ladders to cover the remaining length of the heap, we can continue.
    heap.enqueue(delta);

    // As mentioned, the heap stores how many jumps that consume ladders or bricks at this point.  Because ladders
    // can represent any number of bricks, this means that we "buy" elements in the heap, or that the heap may have
    // an ambient size.
    //
    // For example, normally each element pushed onto the heap must be popped off.  However, if we have 3 ladders,
    // the min heap is permitted to contain 3 elements that don't need to be popped off.  These 3 elements are the
    // biggest deltas, and we use ladders to traverse them.
    //
    // At this moment, we don't know which 3 elements will be the biggest, but we just treat 3 as the "floor" size of
    // the heap and continue if we haven't met this floor.
    if (ladders >= heap.size()) {
      // Assume that this delta, at height i, will be claimed by a ladder.
      continue;
    }

    // If, at this point, we've run out of ladders, we'll need to pop off the smallest element and use bricks to cross
    // the delta.  This will cause the largest element to be different, but the ladder doesn't care how big the
    // largest element is.
    const smallest = heap.dequeue();
    bricks -= smallest;

    // If we've run out of bricks, it's okay.  But if we've gone negative into bricks, we can't cross this gap and
    // we return i (since we can't get to i + 1).
    if (bricks < 0) {
      return i;
    }
  }

  // There's no building at heights.length; the last building is at heights.length - 1.
  return heights.length - 1;
}

File: src/leetcode/heap/furthest_building_you_can_reach.py
# DIFFICULTY: Medium
#
# You are given an integer array heights representing the heights of buildings, some bricks, and some ladders.
#
# You start your journey from building 0 and move to the next building by possibly using bricks or ladders.
#
# While moving from building i to building i+1 (0-indexed),
#
# - If the current building's height is greater than or equal to the next building's height, you do not need a ladder
#   or bricks.
# - If the current building's height is less than the next building's height, you can either use one ladder or
#   (h[i+1] - h[i]) bricks.
#
# Return the furthest building index (0-indexed) you can reach if you use the given ladders and bricks optimally.
#
# See https://leetcode.com/problems/furthest-building-you-can-reach
from heapq import heappop, heappush


class Solution:
    def furthestBuilding(self, heights: list[int], bricks: int, ladders: int) -> int:
        """
        SOLUTION:

        You can't do this problem using sliding window or two pointer technique easily.

        Sliding window requires finding some value in a fixed size window.  Here we don't know what the window size will
        be.

        Two pointer technique can solve this problem but not optimally.  We need to dynamically adjust our resources of
        bricks or ladder at each step.

        A greedy algorithm would be better suited to solve this problem.

        COMPLEXITY:

        Time complexity is O(n log n) time where n is number of buildings.

        Space complexity is O(n) in worst case.
        """
        heap: list[int] = []

        for i in range(len(heights) - 1):
            # Calculate the delta that we need to "jump" across.
            delta = heights[i + 1] - heights[i]

            # A negative or zero height jump can be traversed by jumping across or down.
            if delta <= 0:
                continue

            # A non-negative delta means we must use either bricks or a ladder to jump across.
            heappush(heap, delta)

            # Because the ladders can traverse any height, we'll "save" the ladders here to use on the the biggest
            # deltas.  As long as we have ladders remaining we can maintain a heap size of deltas that we'll
            # eventually use ladders on.
            if len(heap) <= ladders:
                continue

            # If we have more elements on the heap than we do ladders, then SOME of these deltas need to be traversed
            # using bricks instead of ladders.  Let's pop the smallest delta out and use bricks on it.
            smallest = heappop(heap)
            bricks -= smallest

            # If we've run out of bricks, that's it.  We can can only reach the building at position i.
            if bricks < 0:
                return i

        # If we got here, that means we can get to the very last building!
        return len(heights) - 1
